Simple Payments: Improve price formatting

sirreal opened the pull request:
With the Simple Payments block work, it was decided that the client sending a formatted price meta was not an appropriate way to store the price to be displayed.

This PR adds improved price formatters largely based on some used in WordPress.com and renders the price for display when the shortcode is processed.

Extracted from #10402

##### Some examples

###### Yen / es_ES locale
![screen shot 2018-11-07 at 13 02 03](https://user-images.githubusercontent.com/841763/48130610-7e2dd880-e28d-11e8-8da8-6af60905f806.png)

###### Yen / en_US locale
![screen shot 2018-11-07 at 13 02 53](https://user-images.githubusercontent.com/841763/48130614-8128c900-e28d-11e8-9ec7-d964d3caac80.png)

###### USD / en_US
![screen shot 2018-11-07 at 13 05 52](https://user-images.githubusercontent.com/841763/48130737-e67cba00-e28d-11e8-8e62-42346d41b403.png)

###### GBP / en_US (hovering)

![screen shot 2018-11-07 at 13 07 17](https://user-images.githubusercontent.com/841763/48130790-175cef00-e28e-11e8-9984-bbfe031ce006.png)


#### Changes proposed in this Pull Request:

* Add improved Simple Payments price formatting in the shortcode.

#### Testing instructions:

* With a Jetpack site and a Premium or greater plan, add a Simple Payments button.
* Verify that a variety of currencies display well when the Simple Payment is rendered on the frontend.


#### Proposed changelog entry for your changes:

* Improve Simple Payments frontend formatting.

sirreal committed:
Add get_currency to Simple Payments

sirreal committed:
Update format_price

simison commented on modules/simple-payments/simple-payments.php:
```diff
@@ -241,7 +239,7 @@ function output_shortcode( $data ) {
 			$data['description'],
 			esc_attr( "${css_prefix}-price" ),
-			esc_html( $data['price'] ),
+			$data['price'], // Escaped by format_price
```
Should we move `esc_html` here just so that we do it as late as possible before output, to be clear that we're really doing it?

sirreal replied:
That's an important question. I followed some existing patterns that produce appropriately escaped html for this bit. It results in something like `<abbr title="United States Dollars">$</abbr> 42.99`. If we escape here, that doesn't turn out well 🙂 

I believe that's a change from the existing implementation and I'm not sure whether it's desired. If we just want to produce `$ 42.99`, we can continue to escape here and everything's simpler.

@thomasguillot Thoughts?

simison replied:
Oh right, that makes sense then. I didn't test run this PR yet, just followed my gut feeling of _better escape late_.

I'll leave it to @thomasguillot if `$` is enough or if it needs `<abbr>` :-)

simison replied:
>Also in that case it's a misuse of the `abbr` tag because `<abbr title="United States Dollars">` should be around `USD` and not `$`. Anyway... like I said, it's not needed 😊

Valid point! :thumbsup:

> the sign for SGD should be S$, just like C$ for CAD, etc...

Gotcha, we can fix this.

simison replied:
Singapore dollar $ follow-ups here: https://github.com/Automattic/wp-calypso/pull/28380

simison commented on modules/simple-payments/simple-payments.php:
```diff
@@ -377,5 +404,173 @@ function setup_cpts() {
+	 * @return ?array               Currency object or null if not found.
+	 */
+	private static function get_currency( $the_currency ) {
+		$currencies = array(
```
Looks like we're missing `MYR` (Malaysian ringgit) here and `INR` (India rupee) is extraneous.

simison replied:
Resolved in https://github.com/Automattic/jetpack/pull/10555/commits/2ef9d9065893b33403dbb527dda61cae28f3188b and https://github.com/Automattic/jetpack/pull/10555/commits/8738f686c4ddc22e1767deb630a22afd3ff3e433

simison committed:
Add Malaysian ringgit

Since it's on PayPal's supported list:
https://developer.paypal.com/docs/integration/direct/rest/currency-codes/

Picked values from:
https://github.com/Automattic/wp-calypso/blob/474b1ecbfb0f7002fd0c00da2aac48fe2abacb8a/client/lib/format-currency/currencies.js#L585-L590
and cross-checked with:
https://en.wikipedia.org/wiki/Malaysian_ringgit

simison committed:
Remove Indian Rupees

simison commented:
Noting that in Calypso we [support also formatting grouping and decimal symbols](https://github.com/Automattic/wp-calypso/blob/37830bdc69f2bddfac1746e5bb1b9a120510d509/client/lib/format-currency/currencies.js) which we don't do in this PR.

sirreal commented:
> Noting that in Calypso we support also formatting grouping and decimal symbols which we don't do in this PR.

That's a good point. This relies on the locale (`number_format_i18n`) rather than anything currency-specific, except in the case of USD ($). I'm not sure what's more correct.

@Automattic/i18n Do you have thoughts on whether number formatting of currencies should be currency- or locale-specific?

simison commented on modules/simple-payments/simple-payments.php:
```diff
@@ -377,5 +404,173 @@ function setup_cpts() {
+	 *   format  => string sprintf format with placeholders `%1$s`: Symbol `%2$s`: Price.
+	 *   symbol  => string Symbol string
+	 *   desc    => string Text description of currency
+	 *   decimal => int    Number of decimal places
```
FYI I confirmed that all the decimals match with [Calypso side](https://github.com/Automattic/wp-calypso/blob/37830bdc69f2bddfac1746e5bb1b9a120510d509/client/lib/format-currency/currencies.js)

simison commented:
> Automattic/i18n Do you have thoughts on whether number formatting of currencies should be currency- or locale-specific?

I'd suggest we can merge without those and loop back to add them if needed.

simison commented:
There's a currency list also at widged code at `modules/widgets/simple-payments.php` — in perhaps another PR we should consodidate these to one place.

https://github.com/Automattic/jetpack/blob/110920b0527d2e7f6781ea9b34a8943eda72458b/modules/widgets/simple-payments.php#L16-L42

It's used here:
https://github.com/Automattic/jetpack/blob/110920b0527d2e7f6781ea9b34a8943eda72458b/modules/widgets/simple-payments/form.php#L119-L123

simison reviewed (dismissed):
Tests well:

- created button using branch from https://github.com/Automattic/jetpack/pull/10555 and simple payments block, which doesn't send `formattedPrice` meta.
- Looked at the payment button at theme side with this branch:
    <img width="238" alt="image" src="https://user-images.githubusercontent.com/87168/48192759-5c962500-e351-11e8-9c75-66d1ad2aad06.png">


Let's just remove `abbr` as per convo and 🚢 
https://github.com/Automattic/jetpack/pull/10555/files#r231797036

In later PR we should use this list for the widget (https://github.com/Automattic/jetpack/pull/10555#issuecomment-436944218) and possibly check back about number formatting (https://github.com/Automattic/jetpack/pull/10555#issuecomment-436935151). Neither issue is blocking us from shipping the Gutenblock.

sirreal committed:
Simplify price formatting and escaping without abbr

sirreal commented:
I've simplified the formatting to remove the `<abbr>` and simplifying the HTML escapint (once and late – matches the rest of the output).

This should be ready to ship!

tyxla commented on modules/simple-payments/simple-payments.php:
```diff
@@ -377,5 +398,173 @@ function setup_cpts() {
+			'BRL' => array(
+				'format'  => '%1$s%2$s', // 1: Symbol 2: currency value
+				'symbol'  => 'R$',
+				'desc'    => _x( 'Brazilian real', 'currency', 'jetpack' ),
```
Is it intended to have some currencies with sentence case vs some with capital case (USD for example)?

sirreal replied:
Descriptions are now unused and removed.

sirreal committed:
Remove unused currency descriptions

simison reviewed (approved):
Re-tested with latest changes. Works well :ship:

Tested with USD:
<img width="236" alt="image" src="https://user-images.githubusercontent.com/87168/48193889-358d2280-e354-11e8-83e8-5a2c2847ce14.png">

jeherve reviewed (approved):
This seems to work well on my ends (tested with EUR, USD, and HUF). Merge when ready! 🚢

jeherve commented:
Cherry-picked to `branch-6.7` in 0c1803f60770f36600034f59ff31f204b57435fc
